plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/weighers/KindWeigher.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+package org.jetbrains.kotlin.idea.completion.weighers
+
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementWeigher
+import com.intellij.openapi.util.Key
+import org.jetbrains.kotlin.idea.completion.KeywordLookupObject
+import org.jetbrains.kotlin.idea.completion.contributors.keywords.ReturnKeywordHandler.isReturnAtHighlyLikelyPosition
+import org.jetbrains.kotlin.idea.completion.lookups.KotlinCallableLookupObject
+import org.jetbrains.kotlin.idea.completion.lookups.factories.PackagePartLookupObject
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtEnumEntrySymbol
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtSymbol
+import org.jetbrains.kotlin.psi.UserDataProperty
+
+internal object KindWeigher {
+    private enum class Weight {
+        RETURN_AT_HIGHLY_LIKELY_POSITION,
+        ENUM_MEMBER,
+        CALLABLE,
+        KEYWORD,
+        DEFAULT,
+        PACKAGES
+    }
+
+    const val WEIGHER_ID = "kotlin.kind"
+
+    private var LookupElement.isEnumEntry by UserDataProperty(Key<Boolean>("KOTLIN_KIND_WEIGHER_IS_ENUM"))
```

Comment:
Here I would advise the same things as for the `deprecated` - explicitly specified type and `NotNullableUserDataProperty` class instead of `UserDataProperty`